### PR TITLE
ChangeLog 3.x to 4.x - Closes #560

### LINKED ISSUE #560: Update developer changelog
## Which content is outdated?
The developer changelog refers to changes form SDK 2.x to 3.x

## What should be the new content?
The changelog should reflect the changes from SDK 3.x to SDK 4.x

### PR DESCRIPTION
**Pull request by Tschakki**
Closes #560

### CODE REVIEW

File: modules/ROOT/pages/references/changelog.adoc [hunk 1]
```diff
@@ -1,100 +1,344 @@
...
+----
+--
+=====
 
-The new parameter `networkIdentifier` is introduced with Lisk SDK version 3.x.
+=== Update: Vote transaction
 
-It is now required to provide the `networkIdentifier` of the specific network when creating a transaction.
-This eliminates the possibility of transaction replays on different networks.
+The voting process has changes as shown in the code snippet below.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The voting process now has certain changes which can be seen in the code snippet below.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 2]
```diff
@@ -1,100 +1,344 @@
...
-* The Community identifier specified in the genesis block. (Example: {url_github_devnet_genesis_comid}[Devnet genesis block Community Identifier^]).
+In v4.0.0 xref:{url_bigint}[BigInt] is used to handle large numbers, like account balances and transfer amounts.
 
+BigInt is a built-in JavaScript object and replaces the previously used `BigNum` object.
+
+IMPORTANT: The balances of accounts are now saved as **BigInt**(previously `string`) in the Lisk SDK v4.0.0.
+
+== applyAsset() becomes asynchronous
+
+The `applyAsset()` function of custom transaction has become asynchronous, so you need to adjust the function in your implementation to something like this:
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The `applyAsset()` function of custom transaction has become asynchronous, so it is necessary to adjust the function in your implementation to something similar as shown below:
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 3]
```diff
@@ -1,100 +1,344 @@
...
+--
+=====
 
-const networkIdentifier = getNetworkIdentifier(
-    "23ce0366ef0a14a91e5fd4b1591fc880ffbef9d988ff8bebf8f3666b0c09597d",
-    "Lisk",
-);
+== The store returns account as Account instances
+
+The accounts that are returned from the xref:{url_guides_tx_store}[store] in transactions are now returned as instances of the Account class instead of pure objects.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The accounts that are returned from the xref:{url_guides_tx_store}[store] in transactions are now returned as instances of the account class instead of pure objects.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 4]
```diff
@@ -1,100 +1,344 @@
...
+
+Because of this, it is recommended to mutate the received account instance directly.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
Due to this, it is recommended to mutate the received account instance directly.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 5]
```diff
@@ -1,100 +1,344 @@
...
+}
+----
+--
+=====
+
+== Dynamic fees
+
+Static fees are replaced with dynamic fees (see: {url_github_lip13}[LIP 13: Replace static fee system by dynamic fee system^]).
+
+The properties of custom transactions can now be updated like so:
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The properties of custom transactions can now be updated as shown below:
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 6]
```diff
@@ -1,100 +1,344 @@
...
+
+* Removed: The static `FEE` property in a custom transaction.
+* Added: The static property `MIN_FEE_PER_BYTE`.
+It defines the minimum amount of tokens that need to paid per Byte of a particular transaction type.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
This defines the minimum amount of tokens that need to paid per byte of a particular transaction type.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 7]
```diff
@@ -1,100 +1,344 @@
...
+It defaults to `1000`, but can be overwritten as desired.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
This defaults to `1000`, but can be overwritten as desired.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 8]
```diff
@@ -1,100 +1,344 @@
...
+* Added: The static property `NAME_FEE`.
+It defines an absolute amount of tokens that always need to be paid as transaction fee, in addition to the fees generated by the `MIN_FEE_PER_BYTE` property.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
This defines a specific amount of tokens that always need to be paid as a transaction fee, in addition to the fees generated by the `MIN_FEE_PER_BYTE` property.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 9]
```diff
@@ -1,100 +1,344 @@
...
+It defaults to `0`, but can be overwritten as desired.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
This defaults to `0`, but can be overwritten as desired.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 10]
```diff
@@ -1,100 +1,344 @@
...
-const getTimestamp = () => {
-    // check config file or curl localhost:4000/api/node/constants to verify your epoc time
-    const millisSinceEpoc = Date.now() - Date.parse(EPOCH_TIME);
-    const inSeconds = ((millisSinceEpoc) / 1000).toFixed(0);
-    return  parseInt(inSeconds);
-};
+== Nonce replaces the timestamp in transactions
 
-const tx = new HelloTransaction({
+The timestamp which has been a required property previously has been replaced by a nonce.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The timestamp which was previously a required property has been replaced by a nonce.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 11]
```diff
@@ -1,100 +1,344 @@
...
+        hello: 'Hello SDK 4.0',
     },
-    networkIdentifier: networkIdentifier,
-    timestamp: getTimestamp(),
+    fee: '116000',
+    nonce: '0',
 });
+----
+
+Nonces prevent transaction replays and enable users to invalidate transactions, in case of need.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
Nonces prevent transaction replays and enable users to invalidate transactions, if so required
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 12]
```diff
@@ -1,100 +1,344 @@
...
+const passphrase = 'My secret passphrase';
+
+helloTransaction.sign(networkIdentifier,passphrase);
+----
+
+== Accounts must maintain a minimum balance
+
+{url_github_lip25}[LIP 25: Introduce minimum balance requirement for accounts^] introduces a new constant `minBbalance`, which is defining a minimum balance for all accounts in the network.
+
+If the minimum balance for accounts is set to a reasonable value, it prevents the network from certain spam attacks, like creating a lot of new accounts with basically no balance.
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
If the minimum balance for accounts is set to a reasonable value, it prevents the network from certain spam attacks, such as the creation of many new accounts with zero balance.
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 13]
```diff
@@ -106,246 +350,183 @@ See the new devnet config: https://github.com/LiskHQ/lisk-sdk/blob/v3.0.2/sdk/sr
...
+  "custom" : true,
+  "field" : true
+}`|object
+|**balance** +
+__required__|The current balance of the account in Beddows. +
+**Example** : `"1081560729258"`|string
+|**delegate** +
+__required__||<<_delegate,Delegate>>
+|**isDelegate** +
+__optional__|The value indicating if the account is a deleage or not.|boolean
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
__optional__|The value indicating if the account is a delegate or not.|boolean
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 14]
```diff
@@ -106,246 +350,183 @@ See the new devnet config: https://github.com/LiskHQ/lisk-sdk/blob/v3.0.2/sdk/sr
...
+|**keys** +
+__required__||<<_multisignatureasset,MultisignatureAsset>>
+|**missedBlocks** +
+__optional__|Total number of blocks that the delegate has missed. +
+**Example** : `427`|integer
+|**nonce** +
+__required__|The current nonce associated to account for transaction processing. +
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
__required__|The current nonce associated to an account for transaction processing. +
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 15]
```diff
@@ -106,246 +350,183 @@ See the new devnet config: https://github.com/LiskHQ/lisk-sdk/blob/v3.0.2/sdk/sr
...
-__optional__|The version of Lisk Core that the peer node runs on. +
-**Example** : `"v0.8.0"`|string (version)
-|**wsPort** +
-__required__|The port the peer node uses for websocket connections, e.g. P2P broadcasts. +
-**Minimum value** : `1` +
-**Maximum value** : `65535` +
-**Example** : `8001`|integer (int32)
+**publicKey** +
+__required__|The public key is derived from the private key of the owner of the account.
+It can be used to validate that the private key belongs to the owner, but does not provide access to the owners private key. +
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
This can be used to validate that the private key belongs to the owner, but does not provide access to the owners private key. +
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 16]
```diff
@@ -106,246 +350,183 @@ See the new devnet config: https://github.com/LiskHQ/lisk-sdk/blob/v3.0.2/sdk/sr
...
+**Example** : `"968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b"`|string (publicKey)
+| {set:cellbgcolor:red}**secondPublicKey** +
+__optional__|The second public key is derived from the second private key of an account, if the owner activated a second passphrase for her/his account. +
+It can be used to validate that the private key belongs to the owner, but does not provide access to the owners private key. +
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
This can be used to validate that the private key belongs to the owner, but does not provide access to the owners private key. +
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 17]
```diff
@@ -481,291 +660,431 @@ __required__|The delegates' username.
...
+__required__|Account balance.
+Amount of Lisk the delegate account owns. +
+**Example** : `"1081560729258"`|string
+|**publicKey** +
+__optional__|The public key of the delegate. +
+**Example** : `"968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b"`|string (publicKey)
+|**username** +
+__required__|The delegates username.
+A delegate chooses the username by registering a delegate on the Lisk network.
+It is unique and cannot be changed later. +
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
It is unique and cannot be changed. +
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 18]
```diff
@@ -481,291 +660,431 @@ __required__|The delegates' username.
...
+
+[options="header", cols=",,"]
+|===
+|Name|Description|Schema
+|{set:cellbgcolor:red}**publicKey** +
+__required__|Public key of the account that intends to sign the multisignature transaction. +
+**Example** : `"2ca9a7143fc721fdc540fef893b27e8d648d2288efa61e56264edf01a2c23079"`|string (publicKey)
+|**signature** +
+__required__|Signature to sign the transaction.
+The signature can be generated locally, either by using Lisk Commander or with Lisk elements. +
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The signature can be generated locally, either by using Lisk Commander or with Lisk Elements. +
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 19]
```diff
@@ -776,56 +1095,252 @@ By forging a new block on a chain, all former blocks and their contained transac
...
+|**nonce** +
+__required__|Unique sequence of number per account. +
+**Example** : `"0"`|string
+|{set:cellbgcolor!}
+**senderPublicKey** +
+__required__|The public key of the senders' account. +
+**Example** : `"2ca9a7143fc721fdc540fef893b27e8d648d2288efa61e56264edf01a2c23079"`|string (publicKey)
+|{set:cellbgcolor:red}
+**signSignature** +
+__optional__|Contains the second signature, if the transaction is sent from an account with second passphrase activated. +
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
__optional__|Contains the second signature, if the transaction is sent from an account with a second passphrase activated. +
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 20]
```diff
@@ -776,56 +1095,252 @@ By forging a new block on a chain, all former blocks and their contained transac
...
+|**Query**|**address** +
+__optional__|Address of an account.|string (address)|
+|**Query**|**limit** +
+__optional__|Limit applied to results.|integer (int32)|`10`
+|**Query**|**offset** +
+__optional__|Offset value for results.|integer (int32)|`0`
+|**Query**|**publicKey** +
+__optional__|Public key to query.|string (publicKey)|
+|{set:cellbgcolor:red}**Query**|**secondPublicKey** +
+__optional__|Second public key to query|string (publicKey)|
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
__optional__|Second public key to query.|string (publicKey)|
```

File: modules/ROOT/pages/references/changelog.adoc [hunk 21]
```diff
@@ -776,56 +1095,252 @@ By forging a new block on a chain, all former blocks and their contained transac
...
```
Comment (chris529) [inline modules/ROOT/pages/references/changelog.adoc]:
```suggestion
__optional__|Second public key to query.|string (publicKey)|
```

**Review by chris529 — COMMENTED** [review]
Just a few minor points 👍

File: modules/ROOT/pages/references/changelog.adoc (hunk 3 above)
Comment (Tschakki) [reply modules/ROOT/pages/references/changelog.adoc]:
```suggestion
The accounts that are returned from the xref:{url_guides_tx_store}[store] in transactions are now returned as instances of the `Account` class instead of pure objects.
```

**Review by chris529 — DISMISSED** [review]
LGTM!

File: modules/ROOT/pages/guides/app-development/custom-transactions.adoc [hunk 22]
```diff
@@ -238,7 +238,7 @@ Undoes all of the changes to the accounts applied by the `applyAsset` step.
 
 === Additional methods
 
-To increase your application’s performance, the following functions should be overidden: `verifyAgainstTransactions`, `assetFromSync`, `fromSync`.
+To increase your application’s performance, the following functions should be overidden: `MIN_FEE_PER_BYTE`, `NAME_FEE`, `verifyAgainstTransactions`, `assetFromSync`, `fromSync`.
```
Comment (shuse2) [inline modules/ROOT/pages/guides/app-development/custom-transactions.adoc]: Maybe rephrase this paragraph because fee does not increase performance.
Also, 
> verifyAgainstTransactions`, `assetFromSync`, `fromSync`. no longer exist and requried.

File: modules/ROOT/pages/references/changelog.adoc [hunk 23]
```diff
@@ -1,100 +1,344 @@
...
+        mandatoryKeys: [
+            '9d3058175acab969f41ad9b86f7a2926c74258670fe56b37c429c01fca9f2f0f',
+            '141b16ac8d5bd150f16b1caa08f689057ca4c4434445e56661831f4e671b7c0a',
+            '3ff32442bb6da7d60c1b7752b24e6467813c9b698e0f278d48c43580da972135',
+        ],
+        optionalKeys: [],
+        numberOfSignatures: 3,
+    },
+    nonce: '2',
+    fee: '250000',
```
Comment (shuse2) [inline modules/ROOT/pages/references/changelog.adoc:123]: I think as general practice, senderPublicKey should be included when instantiating
because with the new multi-sig, it's possible that senderPublicKey does not associate with the signature

File: modules/ROOT/pages/references/changelog.adoc [hunk 24]
```diff
@@ -1,100 +1,344 @@
...
+});
+console.log(tx);
+/*
+MultisignatureTransaction {
```
Comment (shuse2) [inline modules/ROOT/pages/references/changelog.adoc]: Do we need to show this class instance?

**Review by shuse2 — CHANGES_REQUESTED** [review]
Assuming that the API doc comes from swagger, some small minor points.
Otherwise LGTM

File: modules/ROOT/pages/references/changelog.adoc (hunk 23 above)
Comment (Tschakki) [reply modules/ROOT/pages/references/changelog.adoc:123]: Imho it is not necessary to provide the senderPublicKey in this step, because it gets added automatically once the transaction is signed with the `sign()` method.

File: modules/ROOT/pages/references/changelog.adoc (hunk 24 above)
Comment (Tschakki) [reply modules/ROOT/pages/references/changelog.adoc]: Not necessarily.. I removed it now 👍

File: modules/ROOT/pages/references/changelog.adoc (hunk 23 above)
Comment (shuse2) [reply modules/ROOT/pages/references/changelog.adoc:123]: it does now for the single signature account, but for multi-signature account it will be required to put the sender public key. (sender public key cannot be derived from the passphrase anymore

Comment (Tschakki) [reply modules/ROOT/pages/references/changelog.adoc:76]: What does this mean from UX perspective? Can it be derived from the Address?
If so, do I need to add my address/public key to every transaction before I can sign it with my passphrase?
Sounds not like an optimal solution to me UX-wise. 
Can you provide with a good UX solution for this? 
Otherwise I would keep it like this for now, as for 4.x this is still absolutely valid.

Comment (Tschakki) [reply modules/ROOT/pages/references/changelog.adoc:123]: I see! I will add a new section about creating multisig transactions in the changelog, and also in the app-development guides, to cover the topic. I wouldn't update the general examples, to keep it simple.

File: modules/ROOT/pages/guides/app-development/custom-transactions.adoc [hunk 25]
```diff
@@ -236,9 +236,9 @@ async undoAsset(store: StateStore): ReadonlyArray<TransactionError>
 The inversion of the `applyAsset` method.
 Undoes all of the changes to the accounts applied by the `applyAsset` step.
 
-=== Additional methods
+=== Additional methods & properties
 
-To increase your application’s performance, the following functions should be overidden: `MIN_FEE_PER_BYTE`, `NAME_FEE`, `verifyAgainstTransactions`, `assetFromSync`, `fromSync`.
+It's recommended to override the values of `MIN_FEE_PER_BYTE`, `NAME_FEE` properties of a custom transaction depending on the use case.
```
Comment (shuse2) [inline modules/ROOT/pages/guides/app-development/custom-transactions.adoc]: im not sure we should recommend this? it's more like if u have solid reason to change, then it can be changed, but otherwise it should remain as default

**Review by shuse2 — APPROVED** [review]
